File: python/flox_py/portfolio_risk.py
"""Portfolio-level risk aggregator for multi-strategy setups.

Per-strategy risk hooks are first-class in flox; this module is the
layer above. It aggregates PnL, gross / net exposure, and trade
counts across N registered strategies, applies portfolio-level rules
(max drawdown, max gross exposure, max position concentration),
and publishes a single kill-switch signal back to the engine.

Phase 1 is single-process and in-memory. The aggregator now lives in
the C++ engine; this module is the Python surface, with the same
dataclass shape it always had.
"""
from __future__ import annotations

import threading
from dataclasses import dataclass, field
from typing import Any, Callable, Dict, List, Optional


_FIELD_TO_BIT: Dict[str, int] = {
    "realized_pnl": 1 << 0,
    "unrealized_pnl": 1 << 1,
    "fees": 1 << 2,
    "gross_exposure": 1 << 3,
    "net_exposure": 1 << 4,
    "trade_count": 1 << 5,
}


@dataclass
class StrategyAccount:
    name: str
    realized_pnl: float = 0.0
    unrealized_pnl: float = 0.0
    fees: float = 0.0
    gross_exposure: float = 0.0
    net_exposure: float = 0.0
    trade_count: int = 0

    @property
    def daily_pnl(self) -> float:
        return self.realized_pnl + self.unrealized_pnl + self.fees


@dataclass
class RiskRules:
    max_drawdown_pct: Optional[float] = None
    max_daily_loss: Optional[float] = None
    max_gross_exposure: Optional[float] = None
    max_concentration_pct: Optional[float] = None


@dataclass
class Breach:
    rule: str
    value: float
    limit: float
    detail: str = ""


@dataclass
class PortfolioSnapshot:
    total_realized_pnl: float
    total_unrealized_pnl: float
    total_fees: float
    total_daily_pnl: float
    total_gross_exposure: float
    total_net_exposure: float
    peak_equity: float
    drawdown_pct: float
    accounts: List[StrategyAccount]
    kill_switch_active: bool
    breaches: List[Breach] = field(default_factory=list)

    def to_dict(self) -> dict:
        return {
            "total_realized_pnl": self.total_realized_pnl,
            "total_unrealized_pnl": self.total_unrealized_pnl,
            "total_fees": self.total_fees,
            "total_daily_pnl": self.total_daily_pnl,
            "total_gross_exposure": self.total_gross_exposure,
            "total_net_exposure": self.total_net_exposure,
            "peak_equity": self.peak_equity,
            "drawdown_pct": self.drawdown_pct,
            "accounts": [
                {
                    "name": a.name,
                    "realized_pnl": a.realized_pnl,
                    "unrealized_pnl": a.unrealized_pnl,
                    "fees": a.fees,
                    "daily_pnl": a.daily_pnl,
                    "gross_exposure": a.gross_exposure,
                    "net_exposure": a.net_exposure,
                    "trade_count": a.trade_count,
                }
                for a in self.accounts
            ],
            "kill_switch_active": self.kill_switch_active,
            "breaches": [
                {"rule": b.rule, "value": b.value,
                 "limit": b.limit, "detail": b.detail}
                for b in self.breaches
            ],
        }


KillSwitchCallback = Callable[[List[Breach]], None]


def _account_from_dict(d: dict) -> StrategyAccount:
    return StrategyAccount(
        name=d["name"],
        realized_pnl=float(d["realized_pnl"]),
        unrealized_pnl=float(d["unrealized_pnl"]),
        fees=float(d["fees"]),
        gross_exposure=float(d["gross_exposure"]),
        net_exposure=float(d["net_exposure"]),
        trade_count=int(d["trade_count"]),
    )


def _breach_from_dict(d: dict) -> Breach:
    return Breach(
        rule=str(d["rule"]),
        value=float(d["value"]),
        limit=float(d["limit"]),
        detail=str(d.get("detail", "")),
    )


def _build_native_rules(rules: RiskRules) -> Any:
    from flox_py._flox_py import _PortfolioRiskRules  # type: ignore[attr-defined]

    n = _PortfolioRiskRules()
    n.max_drawdown_pct = rules.max_drawdown_pct
    n.max_daily_loss = rules.max_daily_loss
    n.max_gross_exposure = rules.max_gross_exposure
    n.max_concentration_pct = rules.max_concentration_pct
    return n


class PortfolioRiskAggregator:
    """Single-process portfolio risk aggregator. Delegates to the
    C++-backed implementation in ``flox_py._flox_py`` and preserves
    the dataclass-friendly Python surface."""

    def __init__(
        self,
        *,
        rules: Optional[RiskRules] = None,
        initial_equity: float = 0.0,
        on_breach: Optional[KillSwitchCallback] = None,
    ) -> None:
        from flox_py._flox_py import _PortfolioRiskAggregator  # type: ignore[attr-defined]

        self._rules = rules or RiskRules()
        self._on_breach = on_breach
        self._native = _PortfolioRiskAggregator(
            _build_native_rules(self._rules), float(initial_equity),
        )
        self._kill_switch_seen = False
        self._callback_lock = threading.Lock()

    def update(self, name: str, **fields: float) -> None:
        valid = set(_FIELD_TO_BIT.keys())
        unknown = set(fields.keys()) - valid
        if unknown:
            raise AttributeError(
                f"unknown StrategyAccount field {next(iter(unknown))!r}"
            )
        mask = 0
        for k in fields:
            mask |= _FIELD_TO_BIT[k]

        kwargs: Dict[str, float] = {
            "realized_pnl": float(fields.get("realized_pnl", 0.0)),
            "unrealized_pnl": float(fields.get("unrealized_pnl", 0.0)),
            "fees": float(fields.get("fees", 0.0)),
            "gross_exposure": float(fields.get("gross_exposure", 0.0)),
            "net_exposure": float(fields.get("net_exposure", 0.0)),
            "trade_count": int(fields.get("trade_count", 0)),
        }
        self._native.update(name=name, field_mask=mask, **kwargs)
        self._maybe_fire_breach()

    def remove(self, name: str) -> None:
        self._native.remove(name)
        self._maybe_fire_breach()

    def reset_kill_switch(self) -> None:
        self._native.reset_kill_switch()
        with self._callback_lock:
            self._kill_switch_seen = False

    def check_order(
        self,
        *,
        strategy: str,
        notional: float,
        side: str,
    ) -> Optional[Breach]:
        result = self._native.check_order(strategy, float(notional), side)
        if result is None:
            return None
        return _breach_from_dict(result)

    def snapshot(self) -> PortfolioSnapshot:
        d = self._native.snapshot()
        accounts = [_account_from_dict(a) for a in d["accounts"]]
        breaches = [_breach_from_dict(b) for b in d["active_breaches"]]
        return PortfolioSnapshot(
            total_realized_pnl=float(d["total_realized_pnl"]),
            total_unrealized_pnl=float(d["total_unrealized_pnl"]),
            total_fees=float(d["total_fees"]),
            total_daily_pnl=float(d["total_daily_pnl"]),
            total_gross_exposure=float(d["total_gross_exposure"]),
            total_net_exposure=float(d["total_net_exposure"]),
            peak_equity=float(d["peak_equity"]),
            drawdown_pct=float(d["drawdown_pct"]),
            accounts=accounts,
            kill_switch_active=bool(d["kill_switch_active"]),
            breaches=breaches,
        )

    def _maybe_fire_breach(self) -> None:
        if self._on_breach is None:
            return
        snap = self.snapshot()
        if not snap.kill_switch_active or not snap.breaches:
            return
        with self._callback_lock:
            if self._kill_switch_seen:
                return
            self._kill_switch_seen = True
        try:
            self._on_breach(list(snap.breaches))
        except Exception:
            pass


__all__ = [
    "Breach",
    "KillSwitchCallback",
    "PortfolioRiskAggregator",
    "PortfolioSnapshot",
    "RiskRules",
    "StrategyAccount",
]

File: src/risk/portfolio_risk.cpp
/*
 * Flox Engine
 * Developed by FLOX Foundation (https://github.com/FLOX-Foundation)
 *
 * Copyright (c) 2025 FLOX Foundation
 * Licensed under the MIT License. See LICENSE file in the project root for full
 * license information.
 */

#include "flox/risk/portfolio_risk.h"

#include <algorithm>
#include <cmath>
#include <stdexcept>

namespace flox::risk
{

PortfolioRiskAggregator::PortfolioRiskAggregator(RiskRules rules,
                                                 double initial_equity)
    : _rules(std::move(rules)),
      _initial_equity(initial_equity),
      _peak_equity(initial_equity)
{
  if (_rules.max_drawdown_pct.has_value() && !(initial_equity > 0.0))
  {
    throw std::invalid_argument(
        "portfolio risk: max_drawdown_pct needs a positive initial_equity to "
        "measure against");
  }
}

void PortfolioRiskAggregator::update(const std::string& name,
                                     const StrategyAccount& fields,
                                     uint8_t field_mask)
{
  std::lock_guard<std::mutex> guard(_mutex);
  auto& row = _accounts[name];
  row.name = name;
  if (field_mask & field_mask::REALIZED_PNL)
  {
    row.realized_pnl = fields.realized_pnl;
  }
  if (field_mask & field_mask::UNREALIZED_PNL)
  {
    row.unrealized_pnl = fields.unrealized_pnl;
  }
  if (field_mask & field_mask::FEES)
  {
    row.fees = fields.fees;
  }
  if (field_mask & field_mask::GROSS_EXPOSURE)
  {
    row.gross_exposure = fields.gross_exposure;
  }
  if (field_mask & field_mask::NET_EXPOSURE)
  {
    row.net_exposure = fields.net_exposure;
    row.net_exposure_reported = true;
  }
  if (field_mask & field_mask::TRADE_COUNT)
  {
    row.trade_count = fields.trade_count;
  }
  reevaluateLocked();
}

void PortfolioRiskAggregator::remove(const std::string& name)
{
  std::lock_guard<std::mutex> guard(_mutex);
  _accounts.erase(name);
  reevaluateLocked();
}

void PortfolioRiskAggregator::resetKillSwitch()
{
  std::lock_guard<std::mutex> guard(_mutex);
  _kill_switch_active = false;
}

std::optional<Breach> PortfolioRiskAggregator::checkOrder(const std::string& strategy,
                                                          double notional,
                                                          const std::string& side) const
{
  std::lock_guard<std::mutex> guard(_mutex);
  if (_kill_switch_active)
  {
    return Breach{"kill_switch_active", 1.0, 0.0,
                  "portfolio kill switch is engaged"};
  }

  // Direction first, because every limit below honours it. A "reduce" is
  // explicit. A "sell" moves net exposure down and a "buy" up, read against
  // the net this strategy has reported; when it has never reported one, its
  // gross exposure stands in, so a caller who only feeds gross still gets
  // risk-reducing orders through. Those orders have to pass: near a limit is
  // exactly where you need them most, and the earlier code blocked them on
  // every state rule.
  auto it = _accounts.find(strategy);
  const bool haveRow = (it != _accounts.end());
  const double absNotional = std::abs(notional);

  bool reducesExposure = false;
  if (side == "reduce" || side == "REDUCE")
  {
    reducesExposure = true;
  }
  else if (haveRow)
  {
    const double delta = (side == "sell" || side == "SELL") ? -absNotional : absNotional;
    if (it->second.net_exposure_reported)
    {
      const double curNet = it->second.net_exposure;
      reducesExposure = std::abs(curNet + delta) < std::abs(curNet);
    }
    else
    {
      reducesExposure = (delta < 0.0) && (it->second.gross_exposure > 0.0);
    }
  }

  // State-based limits are pre-trade gates too, not just post-fill kill-switch
  // triggers: refuse a new exposure-increasing order once the book is already
  // in breach. The comparison matches breachesLocked() exactly, so the gate
  // and the snapshot never disagree; they used to differ by one unit in the
  // last place, and inside that window orders were rejected while snapshot()
  // showed no breach at all and the operator had nothing to read.
  if (!reducesExposure && _rules.max_daily_loss.has_value())
  {
    const double pnl = totalDailyPnlLocked();
    // The sign of the cap carries no meaning: -10,000 and 10,000 both mean a
    // ceiling of 10,000 on the loss, as breachesLocked() has always read it.
    // Taken literally, a negative cap rejected every order from a flat start
    // and only released once the day was 10,001 in profit, which is not
    // reachable without trading.
    const double cap = std::abs(*_rules.max_daily_loss);
    if (pnl < 0.0 && std::abs(pnl) > cap)
    {
      return Breach{"max_daily_loss", pnl, -cap,
                    "daily PnL " + std::to_string(pnl) + " below allowed loss " + std::to_string(-cap)};
    }
  }
  if (!reducesExposure && _rules.max_drawdown_pct.has_value())
  {
    const double dd = drawdownPctLocked();
    if (dd > *_rules.max_drawdown_pct)
    {
      return Breach{"max_drawdown_pct", dd, *_rules.max_drawdown_pct,
                    "drawdown " + std::to_string(dd) + " over cap " + std::to_string(*_rules.max_drawdown_pct)};
    }
  }

  if (!reducesExposure && _rules.max_gross_exposure.has_value())
  {
    const double cap = *_rules.max_gross_exposure;
    const double proposed = totalGrossLocked() + absNotional;
    if (proposed > cap)
    {
      return Breach{"max_gross_exposure", proposed, cap,
                    "order would push gross exposure to " + std::to_string(proposed) + " (cap " + std::to_string(cap) + ")"};
    }
  }

  // Concentration needs two contributing strategies to mean anything, the
  // same guard breachesLocked() applies. One strategy trading alone is 100%
  // of the book by definition, and without this the gate rejected every
  // opening order a single-strategy desk ever sent while snapshot() reported
  // a clean book.
  if (!reducesExposure && _rules.max_concentration_pct.has_value())
  {
    const double stratGross = (haveRow ? it->second.gross_exposure : 0.0) + absNotional;
    size_t contributors = contributingAccountsLocked();
    if (!haveRow || it->second.gross_exposure <= 0.0)
    {
      ++contributors;  // this order would make the candidate a contributor
    }

    const double total = totalGrossLocked() + absNotional;
    if (contributors >= 2 && total > 0.0)
    {
      const double conc = stratGross / total;
      if (conc > *_rules.max_concentration_pct)
      {
        return Breach{"max_concentration_pct", conc, *_rules.max_concentration_pct,
                      "strategy concentration " + std::to_string(conc) + " over cap " + std::to_string(*_rules.max_concentration_pct)};
      }
    }
  }

  return std::nullopt;
}

PortfolioSnapshot PortfolioRiskAggregator::snapshot() const
{
  std::lock_guard<std::mutex> guard(_mutex);
  return buildSnapshotLocked();
}

double PortfolioRiskAggregator::totalGrossLocked() const
{
  double s = 0.0;
  for (const auto& [_, a] : _accounts)
  {
    s += a.gross_exposure;
  }
  return s;
}

double PortfolioRiskAggregator::totalDailyPnlLocked() const
{
  double s = 0.0;
  for (const auto& [_, a] : _accounts)
  {
    s += a.dailyPnl();
  }
  return s;
}

double PortfolioRiskAggregator::currentEquityLocked() const
{
  return _initial_equity + totalDailyPnlLocked();
}

size_t PortfolioRiskAggregator::contributingAccountsLocked() const
{
  size_t n = 0;
  for (const auto& [_, a] : _accounts)
  {
    if (a.gross_exposure > 0.0)
    {
      ++n;
    }
  }
  return n;
}

double PortfolioRiskAggregator::drawdownPctLocked() const
{
  if (_peak_equity <= 0)
  {
    return 0.0;
  }
  const double cur = currentEquityLocked();
  // Clamped at 1.0. Unclamped, a small positive peak against a large loss
  // printed values like 500001.0 -- "50,000,100%" -- straight into
  // snapshot().drawdown_pct, the breach text and every report downstream.
  return std::clamp((_peak_equity - cur) / _peak_equity, 0.0, 1.0);
}

std::vector<Breach> PortfolioRiskAggregator::breachesLocked() const
{
  std::vector<Breach> out;
  if (_rules.max_drawdown_pct.has_value())
  {
    const double dd = drawdownPctLocked();
    if (dd > *_rules.max_drawdown_pct)
    {
      out.push_back({"max_drawdown_pct", dd, *_rules.max_drawdown_pct,
                     "drawdown " + std::to_string(dd) + " over cap " + std::to_string(*_rules.max_drawdown_pct)});
    }
  }
  if (_rules.max_daily_loss.has_value())
  {
    const double pnl = totalDailyPnlLocked();
    const double cap = std::abs(*_rules.max_daily_loss);
    if (pnl < 0.0 && std::abs(pnl) > cap)
    {
      out.push_back({"max_daily_loss", pnl, -cap,
                     "daily PnL " + std::to_string(pnl) + " below allowed loss " + std::to_string(-cap)});
    }
  }
  if (_rules.max_gross_exposure.has_value())
  {
    const double gross = totalGrossLocked();
    if (gross > *_rules.max_gross_exposure)
    {
      out.push_back({"max_gross_exposure", gross, *_rules.max_gross_exposure,
                     "gross " + std::to_string(gross) + " over cap " + std::to_string(*_rules.max_gross_exposure)});
    }
  }
  // Contributing strategies, not registered rows. Counting rows meant that
  // when one of two strategies closed out and went flat, the other was
  // suddenly holding "100% of gross" and the kill switch halted the whole
  // portfolio -- on a normal flattening, clearable only by hand, and armed
  // again on the next update.
  if (_rules.max_concentration_pct.has_value() && contributingAccountsLocked() >= 2)
  {
    const double total = totalGrossLocked();
    if (total > 0.0)
    {
      double max_share = 0.0;
      std::string max_name;
      for (const auto& [n, a] : _accounts)
      {
        const double share = a.gross_exposure / total;
        if (share > max_share)
        {
          max_share = share;
          max_name = n;
        }
      }
      if (max_share > *_rules.max_concentration_pct)
      {
        out.push_back({"max_concentration_pct", max_share,
                       *_rules.max_concentration_pct,
                       max_name + " holds " + std::to_string(max_share * 100.0) + "% of gross"});
      }
    }
  }
  return out;
}

void PortfolioRiskAggregator::reevaluateLocked()
{
  const double cur = currentEquityLocked();
  if (cur > _peak_equity)
  {
    _peak_equity = cur;
  }
  if (!_kill_switch_active && !breachesLocked().empty())
  {
    _kill_switch_active = true;
  }
}

PortfolioSnapshot PortfolioRiskAggregator::buildSnapshotLocked() const
{
  PortfolioSnapshot s;
  for (const auto& [_, a] : _accounts)
  {
    s.total_realized_pnl += a.realized_pnl;
    s.total_unrealized_pnl += a.unrealized_pnl;
    s.total_fees += a.fees;
    s.total_gross_exposure += a.gross_exposure;
    s.total_net_exposure += a.net_exposure;
    s.total_trade_count += a.trade_count;
    s.accounts.push_back(static_cast<const StrategyAccount&>(a));
  }
  s.total_daily_pnl = s.total_realized_pnl + s.total_unrealized_pnl + s.total_fees;
  s.current_equity = _initial_equity + s.total_daily_pnl;
  s.peak_equity = _peak_equity;
  s.drawdown_pct = drawdownPctLocked();
  s.kill_switch_active = _kill_switch_active;
  s.active_breaches = breachesLocked();
  return s;
}

}  // namespace flox::risk
